# MODELOS DE PREDICCION DE PEPTIDOS ANTIMICROBIANOS

#### - Edwin Villarraga
#### - Santiago Giraldo
#### - Mauricio
#### - Lesty
#### - Stephany

## <font color='red'>PARTE 1</font>

 ### <font color='blue'>PUNTO 1</font>

## Librerias

In [3]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
%matplotlib inline
from sklearn.preprocessing import StandardScaler
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score
from sklearn.model_selection import train_test_split
from sklearn.linear_model import Lasso
from sklearn.feature_selection import SelectFromModel

## Cargar Datos

In [4]:
#Se cargan los datos negativos
dataN = pd.read_csv('./Datos/DatosNegativos2.csv')

In [5]:
#Se cargan los datos positivos
dataP = pd.read_csv('./Datos/DatosPositivos2.csv')

In [6]:
print("El archivo con datos Negativos tiene {} filas y {} columnas".format(dataN.shape[0],dataN.shape[1]))
print("El archivo con datos Positivos tiene {} filas y {} columnas".format(dataP.shape[0],dataP.shape[1]))

El archivo con datos Negativos tiene 5561 filas y 1761 columnas
El archivo con datos Positivos tiene 8322 filas y 1761 columnas


In [7]:
pd.options.display.max_columns = None
# Modificación de las opciones de pandas para mostrar todas las columnas del dataframe

In [8]:
# Verificación de campos
dataP.head(5)

Unnamed: 0                   sequence  length  molecular_weight  charge  \
0           0        CGGYSGGWKRKRSTSYRCG      19           2109.36   4.859   
1           1           GIMISLMKKLAAHIAK      16           1725.22   3.094   
2           2          FKAWRWAWRMKKLAAPS      17           2133.57   4.995   
3           3       WGWRDIVAGIRKVAAPVLST      20           2195.57   1.996   
4           4  DTHFPICIFCCGCCRQSNCGMCCKT      25           2774.34   0.555   

   charge_density  isoelectric_point     gravy  instability_index  \
0        0.002304          10.056702 -1.310526          73.031579   
1        0.001793          10.302063  0.887500          20.918750   
2        0.002341          12.024475 -0.594118          43.088235   
3        0.000909          10.835022  0.435000          17.010000   
4        0.000200           7.713074  0.392000          31.272000   

   aromaticity  aliphatic_index  boman_index  hydrophobic_ratio       A  \
0     0.157895         0.000000     3.122105           0.105263   0.000   
1     0.000000       140.625000    -0.685000           0.625000  18.750   
2     0.235294        46.470588     1.494706           0.411765  23.529   
3     0.100000       117.000000     0.561500           0.450000  15.000   
4     0.080000        31.200000     0.972000           0.520000   0.000   

        R    N    D       C    E    Q       G     H      I       L       K  \
0  15.789  0.0  0.0  10.526  0.0  0.0  26.316  0.00   0.00   0.000  10.526   
1   0.000  0.0  0.0   0.000  0.0  0.0   6.250  6.25  18.75  12.500  18.750   
2  11.765  0.0  0.0   0.000  0.0  0.0   0.000  0.00   0.00   5.882  17.647   
3  10.000  0.0  5.0   0.000  0.0  0.0  10.000  0.00  10.00   5.000   5.000   
4   4.000  4.0  4.0  32.000  0.0  4.0   8.000  4.00   8.00   0.000   4.000   

        M      F      P       S      T       W       Y     V    AA   AR   AN  \
0   0.000  0.000  0.000  15.789  5.263   5.263  10.526   0.0  0.00  0.0  0.0   
1  12.500  0.000  0.000   6.250  0.000   0.000   0.000   0.0  6.67  0.0  0.0   
2   5.882  5.882  5.882   5.882  0.000  17.647   0.000   0.0  6.25  0.0  0.0   
3   0.000  0.000  5.000   5.000  5.000  10.000   0.000  15.0  5.26  0.0  0.0   
4   4.000  8.000  4.000   4.000  8.000   0.000   0.000   0.0  0.00  0.0  0.0   

    AD   AC   AE   AQ    AG    AH   AI   AL    AK   AM   AF    AP   AS   AT  \
0  0.0  0.0  0.0  0.0  0.00  0.00  0.0  0.0  0.00  0.0  0.0  0.00  0.0  0.0   
1  0.0  0.0  0.0  0.0  0.00  6.67  0.0  0.0  6.67  0.0  0.0  0.00  0.0  0.0   
2  0.0  0.0  0.0  0.0  0.00  0.00  0.0  0.0  0.00  0.0  0.0  6.25  0.0  0.0   
3  0.0  0.0  0.0  0.0  5.26  0.00  0.0  0.0  0.00  0.0  0.0  5.26  0.0  0.0   
4  0.0  0.0  0.0  0.0  0.00  0.00  0.0  0.0  0.00  0.0  0.0  0.00  0.0  0.0   

     AW   AY   AV   RA   RR   RN    RD    RC   RE    RQ   RG   RH   RI   RL  \
0   0.0  0.0  0.0  0.0  0.0  0.0  0.00  5.56  0.0  0.00  0.0  0.0  0.0  0.0   
1   0.0  0.0  0.0  0.0  0.0  0.0  0.00  0.00  0.0  0.00  0.0  0.0  0.0  0.0   
2  12.5  0.0  0.0  0.0  0.0  0.0  0.00  0.00  0.0  0.00  0.0  0.0  0.0  0.0   
3   0.0  0.0  0.0  0.0  0.0  0.0  5.26  0.00  0.0  0.00  0.0  0.0  0.0  0.0   
4   0.0  0.0  0.0  0.0  0.0  0.0  0.00  0.00  0.0  4.17  0.0  0.0  0.0  0.0   

     RK    RM   RF   RP    RS   RT    RW   RY   RV   NA   NR   NN   ND    NC  \
0  5.56  0.00  0.0  0.0  5.56  0.0  0.00  0.0  0.0  0.0  0.0  0.0  0.0  0.00   
1  0.00  0.00  0.0  0.0  0.00  0.0  0.00  0.0  0.0  0.0  0.0  0.0  0.0  0.00   
2  0.00  6.25  0.0  0.0  0.00  0.0  6.25  0.0  0.0  0.0  0.0  0.0  0.0  0.00   
3  5.26  0.00  0.0  0.0  0.00  0.0  0.00  0.0  0.0  0.0  0.0  0.0  0.0  0.00   
4  0.00  0.00  0.0  0.0  0.00  0.0  0.00  0.0  0.0  0.0  0.0  0.0  0.0  4.17   

    NE   NQ   NG   NH   NI   NL   NK   NM   NF   NP   NS   NT   NW   NY   NV  \
0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   
1  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   
2  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0 

In [9]:
# Verificación de campos
dataN.head(2)

Unnamed: 0 sequence  length  molecular_weight  charge  charge_density  \
0           0  MAATRNK       7            790.93   1.996        0.002524   
1           1  MKMDGLR       7            850.06   0.996        0.001172   

   isoelectric_point     gravy  instability_index  aromaticity  \
0          11.000305 -1.014286          77.657143          0.0   
1           8.497986 -0.671429          82.657143          0.0   

   aliphatic_index  boman_index  hydrophobic_ratio       A       R       N  \
0        28.571429     3.387143           0.428571  28.571  14.286  14.286   
1        55.714286     2.661429           0.428571   0.000  14.286   0.000   

        D    C    E    Q       G    H    I       L       K       M    F    P  \
0   0.000  0.0  0.0  0.0   0.000  0.0  0.0   0.000  14.286  14.286  0.0  0.0   
1  14.286  0.0  0.0  0.0  14.286  0.0  0.0  14.286  14.286  28.571  0.0  0.0   

     S       T    W    Y    V     AA   AR   AN   AD   AC   AE   AQ   AG   AH  \
0  0.0  14.286  0.0  0.0  0.0  16.67  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   
1  0.0   0.000  0.0  0.0  0.0   0.00  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   

    AI   AL   AK   AM   AF   AP   AS     AT   AW   AY   AV   RA   RR     RN  \
0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  16.67  0.0  0.0  0.0  0.0  0.0  16.67   
1  0.0  0.0  0.0  0.0  0.0  0.0  0.0   0.00  0.0  0.0  0.0  0.0  0.0   0.00   

    RD   RC   RE   RQ   RG   RH   RI   RL   RK   RM   RF   RP   RS   RT   RW  \
0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   
1  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   

    RY   RV   NA   NR   NN   ND   NC   NE   NQ   NG   NH   NI   NL     NK  \
0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  16.67   
1  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   0.00   

    NM   NF   NP   NS   NT   NW   NY   NV   DA   DR   DN   DD   DC   DE   DQ  \
0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   
1  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   

      DG   DH   DI   DL   DK   DM   DF   DP   DS   DT   DW   DY   DV   CA  \
0   0.00  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   
1  16.67  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   

    CR   CN   CD   CC   CE   CQ   CG   CH   CI   CL   CK   CM   CF   CP   CS  \
0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   
1  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   

    CT   CW   CY   CV   EA   ER   EN   ED   EC   EE   EQ   EG   EH   EI   EL  \
0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   
1  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   

    EK   EM   EF   EP   ES   ET   EW   EY   EV   QA   QR   QN   QD   QC   QE  \
0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   
1  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   

    QQ   QG   QH   QI   QL   QK   QM   QF   QP   QS   QT   QW   QY   QV   GA  \
0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   
1  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   

    GR   GN   GD   GC   GE   GQ   GG   GH   GI     GL   GK   GM   GF   GP  \
0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   0.00  0.0  0.0  0.0  0.0   
1  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  16.67  0.0  0.0  0.0  0.0   

    GS   GT   GW   GY   GV   HA   HR   HN   HD   HC   HE   HQ   HG   HH   HI  \
0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   
1  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   

    HL   HK   HM   HF   HP   HS   HT   HW   HY   HV   IA   IR   IN   ID   IC  \
0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   
1  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0 

No se logra ver en las 1761 columnas la variable respuesta, entonces se adicionará la columna Respuesta=0 en el caso de los negativos y Respuesta=1 en el caso de los positivos.  Luego se unirán los datos en una nuevo Dataframe unificado

In [10]:
dataN['Respuesta']=0

In [11]:
dataP['Respuesta']=1

In [12]:
data = dataN.append(dataP) #se unen los datos de las respuestas positivas y negativas

In [13]:
len_id=data['sequence'].unique

In [14]:
len_id

<bound method Series.unique of 0                        MAATRNK
1                        MKMDGLR
2                        MNIKHYQ
3                        LLWNSIP
4                        MHPHMSY
                  ...           
8317             ILPIIGKILSTIFGK
8318                   YAGAVANDL
8319             KKCFRWQWRMKKLGA
8320          GWAGWLLSPRGSRPSWGP
8321    LRRLYKRVVKLWKRLFRQLRRPVR
Name: sequence, Length: 13883, dtype: object>

la variable sequence es el idenficicador unico de los datos

In [15]:
#Se extrae una lista de las variables que contengan valores nulos
vars_with_na = [var for var in data.columns if data[var].isnull().sum()>1]
vars_with_na

[]

Significa que no hay variables con valores nulos

## Análisis Descriptivo de los Datos

In [16]:
data.shape

(13883, 1762)

In [17]:
# lista de variables numericas
num_vars = [var for var in data.columns if data[var].dtypes != 'O']  #'O' significa python objects
len(num_vars)

1761

Todas las variables son numéricas excepto 1, que es la sequencia, que actua como identificador de la molécula

## Escalamiento

In [18]:
# Se excluye la variable Sequencia
data.drop(['sequence'], axis=1, inplace=True)

In [19]:
data.shape

(13883, 1761)

In [20]:
Xdata = data.drop(['Respuesta'], axis=1, inplace=False) #se crea un dataframe sin la variable respuesta
Ydata = data['Respuesta']
print(Xdata.shape)
print(Ydata.shape)

(13883, 1760)
(13883,)


In [21]:
train_vars = [var for var in Xdata.columns]
all_vars = [var for var in data.columns]

In [22]:
#ajuste del escalador
scaler = StandardScaler()  #se crea una instancia del escalador
scaler.fit(Xdata)

Xdata_std = pd.DataFrame(scaler.transform(Xdata)) #  se ajusta el escalador a los datos de entrenamiento


In [23]:
Xdata_std.head(2)

0         1         2         3         4         5         6  \
0 -1.621112 -1.567719 -1.647289 -0.019215  1.094644  0.894593 -0.945354   
1 -1.620663 -1.567719 -1.575014 -0.353870  0.131734 -0.056603 -0.584815   

          7         8         9       10        11        12        13  \
0  0.870319 -1.102753 -1.266385  1.14450  0.091409  2.584165  0.602939   
1  0.984439 -1.102753 -0.701275  0.73767  0.091409 -0.863028  0.602939   

         14        15       16        17        18        19        20  \
0  2.317285 -0.570322 -0.51604 -0.579515 -0.577311 -0.921235 -0.467583   
1 -0.674234  2.339137 -0.51604 -0.579515 -0.577311  1.029598 -0.467583   

         21        22        23        24       25        26        27  \
0 -0.886105 -1.068507  0.403486  2.636298 -0.78091 -0.666612 -0.850156   
1 -0.886105  0.340929  0.403486  5.992251 -0.78091 -0.666612 -0.850156   

         28        29        30       31        32        33        34  \
0  1.821832 -0.448455 -0.590342 -0.87664  5.865496 -0.226733 -0.170002   
1 -0.751386 -0.448455 -0.590342 -0.87664 -0.296462 -0.226733 -0.170002   

         35       36        37        38       39        40        41  \
0 -0.168532 -0.14897 -0.179553 -0.152015 -0.28813 -0.145561 -0.255284   
1 -0.168532 -0.14897 -0.179553 -0.152015 -0.28813 -0.145561 -0.255284   

         42        43        44        45        46        47         48  \
0 -0.330038 -0.279003 -0.124839 -0.205495 -0.184651 -0.254932  11.044019   
1 -0.330038 -0.279003 -0.124839 -0.205495 -0.184651 -0.254932  -0.209085   

         49        50       51        52        53         54        55  \
0 -0.138379 -0.144547 -0.26498 -0.205787 -0.314051  13.069291 -0.164851   
1 -0.138379 -0.144547 -0.26498 -0.205787 -0.314051  -0.193227 -0.164851   

         56        57        58        59        60        61        62  \
0 -0.196828 -0.157531 -0.164182 -0.244849 -0.133507 -0.237727 -0.280231   
1 -0.196828 -0.157531 -0.164182 -0.244849 -0.133507 -0.237727 -0.280231   

         63        64        65       66       67        68        69  \
0 -0.271849 -0.116979 -0.224233 -0.17804 -0.20059 -0.190032 -0.192644   
1 -0.271849 -0.116979 -0.224233 -0.17804 -0.20059 -0.190032 -0.192644   

         70       71       72        73        74        75        76  \
0 -0.157728 -0.22071 -0.18812 -0.179051 -0.147625 -0.125675 -0.122088   
1 -0.157728 -0.22071 -0.18812 -0.179051 -0.147625 -0.125675 -0.122088   

         77       78        79        80        81        82         83  \
0 -0.128434 -0.13149 -0.170668 -0.117616 -0.169302 -0.216701  14.424998   
1 -0.128434 -0.13149 -0.170668 -0.117616 -0.169302 -0.216701  -0.185468   

         84        85        86        87        88        89        90  \
0 -0.100742 -0.159483 -0.139744 -0.176047 -0.156665 -0.119402 -0.124696   
1 -0.100742 -0.159483 -0.139744 -0.176047 -0.156665 -0.119402 -0.124696   

         91        92        93       94        95        96        97  \
0 -0.169148 -0.176683 -0.147897 -0.11747 -0.111116 -0.092078 -0.129508   
1 -0.169148 -0.176683 -0.147897 -0.11747 -0.111116 -0.092078 -0.129508   

         98         99       100       101       102       103      104  \
0 -0.099722  -0.155138 -0.085028 -0.167015 -0.201946 -0.175331 -0.08996   
1 -0.099722  14.320288 -0.085028 -0.167015 -0.201946 -0.175331 -0.08996   

        105       106       107       108      109       110       111  \
0 -0.155508 -0.137355 -0.152566 -0.157472 -0.09771 -0.116464 -0.169329   
1 -0.155508 -0.137355 -0.152566 -0.157472 -0.09771 -0.116464 -0.169329   

        112       113       114       115       116       117      118  \
0 -0.158739 -0.184789 -0.130264 -0.105332 -0.110221 -0.100515 -0.10669   
1 -0.158739 -0.184789 -0.130264 -0.105332 -0.110221 -0.100515 -0.10669   

        119       120       121       122       123       124       125  \
0 -0.184289 -0.099842 -0.142266 -0.173151 -0.198628 -0.059186 -0.142418   
1 -0.184289 -0.099842 -0.142266 -0.173151 -0.198628 -0

 ### <font color='blue'>PUNTO 2</font>

### Dividir los datos en train y test

In [24]:
Data = pd.concat([Ydata,Xdata],axis=1)

In [25]:
Data.head(2)

Respuesta  Unnamed: 0  length  molecular_weight  charge  charge_density  \
0          0           0       7            790.93   1.996        0.002524   
1          0           1       7            850.06   0.996        0.001172   

   isoelectric_point     gravy  instability_index  aromaticity  \
0          11.000305 -1.014286          77.657143          0.0   
1           8.497986 -0.671429          82.657143          0.0   

   aliphatic_index  boman_index  hydrophobic_ratio       A       R       N  \
0        28.571429     3.387143           0.428571  28.571  14.286  14.286   
1        55.714286     2.661429           0.428571   0.000  14.286   0.000   

        D    C    E    Q       G    H    I       L       K       M    F    P  \
0   0.000  0.0  0.0  0.0   0.000  0.0  0.0   0.000  14.286  14.286  0.0  0.0   
1  14.286  0.0  0.0  0.0  14.286  0.0  0.0  14.286  14.286  28.571  0.0  0.0   

     S       T    W    Y    V     AA   AR   AN   AD   AC   AE   AQ   AG   AH  \
0  0.0  14.286  0.0  0.0  0.0  16.67  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   
1  0.0   0.000  0.0  0.0  0.0   0.00  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   

    AI   AL   AK   AM   AF   AP   AS     AT   AW   AY   AV   RA   RR     RN  \
0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  16.67  0.0  0.0  0.0  0.0  0.0  16.67   
1  0.0  0.0  0.0  0.0  0.0  0.0  0.0   0.00  0.0  0.0  0.0  0.0  0.0   0.00   

    RD   RC   RE   RQ   RG   RH   RI   RL   RK   RM   RF   RP   RS   RT   RW  \
0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   
1  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   

    RY   RV   NA   NR   NN   ND   NC   NE   NQ   NG   NH   NI   NL     NK  \
0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  16.67   
1  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   0.00   

    NM   NF   NP   NS   NT   NW   NY   NV   DA   DR   DN   DD   DC   DE   DQ  \
0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   
1  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   

      DG   DH   DI   DL   DK   DM   DF   DP   DS   DT   DW   DY   DV   CA  \
0   0.00  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   
1  16.67  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   

    CR   CN   CD   CC   CE   CQ   CG   CH   CI   CL   CK   CM   CF   CP   CS  \
0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   
1  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   

    CT   CW   CY   CV   EA   ER   EN   ED   EC   EE   EQ   EG   EH   EI   EL  \
0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   
1  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   

    EK   EM   EF   EP   ES   ET   EW   EY   EV   QA   QR   QN   QD   QC   QE  \
0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   
1  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   

    QQ   QG   QH   QI   QL   QK   QM   QF   QP   QS   QT   QW   QY   QV   GA  \
0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   
1  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   

    GR   GN   GD   GC   GE   GQ   GG   GH   GI     GL   GK   GM   GF   GP  \
0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   0.00  0.0  0.0  0.0  0.0   
1  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  16.67  0.0  0.0  0.0  0.0   

    GS   GT   GW   GY   GV   HA   HR   HN   HD   HC   HE   HQ   HG   HH   HI  \
0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   
1  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   

    HL   HK   HM   HF   HP   HS   HT   HW   HY   HV   IA   IR   IN   ID   IC  \
0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   
1  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0

In [26]:
# Separación de los datos en train y test
# random_state para reproducibilidad

X_train, X_test, y_train, y_test = train_test_split(Data, Data.Respuesta,
                                                    test_size=0.2,
                                                    random_state=0)
X_train.shape, X_test.shape, y_train.shape, y_test.shape

((11106, 1761), (2777, 1761), (11106,), (2777,))

 ### <font color='blue'>PUNTOS 3 y 4</font>

## MODELO 1: Regresión Logística

In [27]:
from sklearn.model_selection import KFold, cross_val_score
import warnings
warnings.filterwarnings('ignore')

In [28]:
kfold = KFold(n_splits=10, random_state=0)
model_kfold = LogisticRegression()
results_kfold = cross_val_score(model_kfold, X_train, y_train, cv=kfold,n_jobs=-1, verbose=1)
print("Accuracy: %.2f%%" % (results_kfold.mean()*100.0)) 

[Parallel(n_jobs=-1)]: Using backend LokyBackend with 8 concurrent workers.
[Parallel(n_jobs=-1)]: Done   6 out of  10 | elapsed:  1.7min remaining:  1.1min


Accuracy: 99.68%


[Parallel(n_jobs=-1)]: Done  10 out of  10 | elapsed:  2.1min finished


## Modelo 2 - Random Forest

In [29]:
from sklearn.ensemble import RandomForestClassifier

In [30]:
kfold = KFold(n_splits=10, random_state=0)
model_kfold = RandomForestClassifier(random_state=0)
results_kfold = cross_val_score(model_kfold, X_train, y_train, cv=kfold, n_jobs=-1, verbose=1)
print("Accuracy: %.2f%%" % (results_kfold.mean()*100.0)) 

[Parallel(n_jobs=-1)]: Using backend LokyBackend with 8 concurrent workers.
[Parallel(n_jobs=-1)]: Done   6 out of  10 | elapsed:    2.8s remaining:    1.8s


Accuracy: 96.88%


[Parallel(n_jobs=-1)]: Done  10 out of  10 | elapsed:    4.4s finished


## Modelo 3 - Máquina de Vectores de Soporte

In [1]:
from sklearn import svm

In [2]:
kfold = KFold(n_splits=2, random_state=0)
model_kfold = svm.SVC(kernel='linear', max_iter=100)
results_kfold = cross_val_score(model_kfold, X_train, y_train, cv=kfold, n_jobs=-1, verbose=1)
print("Accuracy: %.2f%%" % (results_kfold.mean()*100.0)) 

NameError: name 'KFold' is not defined

## Modelo 4 - Adaboost

In [31]:
from sklearn.ensemble import AdaBoostClassifier

In [32]:
kfold = KFold(n_splits=10, random_state=0)
model_kfold = AdaBoostClassifier(n_estimators=100, random_state=0)
results_kfold = cross_val_score(model_kfold, X_train, y_train, cv=kfold, n_jobs=-1, verbose=1)
print("Accuracy: %.2f%%" % (results_kfold.mean()*100.0)) 

[Parallel(n_jobs=-1)]: Using backend LokyBackend with 8 concurrent workers.
[Parallel(n_jobs=-1)]: Done   6 out of  10 | elapsed:    1.9s remaining:    1.2s


Accuracy: 100.00%


[Parallel(n_jobs=-1)]: Done  10 out of  10 | elapsed:    3.1s finished


 ### <font color='blue'>PUNTO 5</font>

## <font color='red'>PARTE 2</font>

## Selección y Reducción de Variables

In [ ]:
#se cargan los datos estandarizados
X_train = pd.read_pickle('./Datos/xtrain.pkl')
X_test = pd.read_pickle('./Datos/xtest.pkl')
y_train = pd.read_pickle('./Datos/ytrain.pkl')
y_test = pd.read_pickle('./Datos/ytest.pkl')

Se usará una regresión lasso,, con una semilla = 0, y la función selectFromModel de sklearn para seleccionar las variables más predictivas.  Entre mayor alpha más reducción de variables

In [ ]:
sel_ = SelectFromModel(Lasso(alpha=0.05, random_state=0)) # remember to set the seed, the random state in this function
sel_.fit(X_train, y_train)

In [ ]:
sel_.get_support()  #para visualizar las varaibles seleccionas

In [ ]:
sum(sel_.get_support())  #se seleccionaron 15 variables significativas según el modelo RL Lasso

In [ ]:
selected_feat = X_train.columns[(sel_.get_support())]

In [ ]:
selected_feat  #caracteristicas seleccionadas

In [ ]:
# algunas estadisticas
print('total variables: {}'.format((X_train.shape[1])))
print('varaibles seleccionadas: {}'.format(len(selected_feat)))
print('Variables excluidas por RL Lasso: {}'.format(
    np.sum(sel_.estimator_.coef_ == 0)))

In [ ]:
selected_feats = X_train.columns[(sel_.estimator_.coef_ != 0).ravel().tolist()]
selected_feats  #otra forma de ver las caracteriticas seleccionadas

In [ ]:
# nse guarda la lista de las caracteristicas seleccionadas
pd.Series(selected_feats).to_csv('./Datos/selected_features.csv', index=False)

## Modelo 1 - Regresión Logística

### Inicialmente con todas las variables

In [ ]:
lr = LogisticRegression(solver='liblinear')
lr.fit(X_train, y_train) #con todas las variables

In [ ]:
X_test.drop(['sequence', 'Respuesta'], axis=1, inplace=True)
y_pred = lr.predict(X_test)

In [ ]:
print("{0:.1%} precisión del modelo".format(accuracy_score(y_test, y_pred))) 


### Ahora con las variables reducidas

In [ ]:
features = pd.read_csv('./Datos/selected_features.csv', header=None)
features = [x for x in features[0]] 

In [ ]:
features

In [ ]:
Xtrain = X_train[features]
Xtest = X_test[features]

In [ ]:
y_train.shape

In [ ]:
#Se guarda la base de datos reducida co las variables seleccionadas
Xtrain.to_csv('./Datos/xtrain.csv', index=False)
Xtest.to_csv('./Datos/xtest.csv', index=False)

In [ ]:
Xtrain.shape

In [ ]:
lr.fit(Xtrain, y_train) #con solo 15 variables

In [ ]:
y_pred = lr.predict(Xtest)

In [ ]:
# Calculates the accuracy on the test set and prints coefficients
acc = accuracy_score(y_test, y_pred)
print("{0:.1%} accuracy on test set.".format(acc)) 

##### Con las 15 variables la precisión del modelo se reduce de 88.2% a 86.3% - pero se redujo el numero de variables de 1760 a 15

## Modelo 2 - Random Forest

In [ ]:
from sklearn.ensemble import RandomForestClassifier

In [ ]:
#carga de datos
Xtrain = pd.read_csv('./Datos/xtrain.csv')
Xtest = pd.read_csv('./Datos/xtest.csv')
y_train = pd.read_pickle('./Datos/ytrain.pkl')
y_test = pd.read_pickle('./Datos/ytest.pkl')

In [ ]:
Xtrain.head(2)

In [ ]:
Xtrain.shape

In [ ]:
y_train.shape

In [ ]:
y_train.head(2)

In [ ]:
clf = RandomForestClassifier(random_state=0)

In [ ]:
clf.fit(Xtrain, y_train)

In [ ]:
y_pred = clf.predict(Xtest)

In [ ]:
# Calculates the accuracy on the test set and prints coefficients
acc = accuracy_score(y_test, y_pred)
print("{0:.1%} accuracy on test set.".format(acc)) 

## Modelo 3 - Máquina de Vectores de Soporte

In [ ]:
from sklearn import svm

In [ ]:
clf = svm.SVC()

In [ ]:
clf.fit(Xtrain, y_train)

In [ ]:
y_pred = clf.predict(Xtest)

In [ ]:
# Calculates the accuracy on the test set and prints coefficients
acc = accuracy_score(y_test, y_pred)
print("{0:.1%} accuracy on test set.".format(acc)) 

## Modelo 4 - Adaboost

In [ ]:
from sklearn.ensemble import AdaBoostClassifier

In [ ]:
>>> clf = AdaBoostClassifier(n_estimators=100, random_state=0)
>>> clf.fit(Xtrain, y_train)

In [ ]:
y_pred = clf.predict(Xtest)

In [ ]:
# Calculates the accuracy on the test set and prints coefficients
acc = accuracy_score(y_test, y_pred)
print("{0:.1%} accuracy on test set.".format(acc)) 